### PR TITLE
8286301: Port JEP 425 to RISC-V

### PR DESCRIPTION
**Pull request by RealFYang**
Hi,

Please review this PR porting JEP 425 (Virtual Threads) to RISC-V.

This is mainly adapted from the work of AArch64 port. Most of the changes lie in RISC-V scope.
Changes to HotSpot shared code are trivial and are always guarded by RISCV64 macro. So this won't
affect the rest of the world in theory.

There exists some differences in frame structure between AArch64 and RISC-V.
For AArch64, we have:
```
enum {
  link_offset           = 0,
  return_addr_offset    = 1,
  sender_sp_offset      = 2
};
```
While for RISC-V, we have:
```
enum {
  link_offset           = -2,
  return_addr_offset    = -1,
  sender_sp_offset      =  0
};
```
So we need adapations in some places where the code relies on value of sender_sp_offset to work.
Note that implementation for Post-call NOPs optimization is not incorporated in this PR as we plan to
evaluate more on its impact on performance.

Testing on Linux-riscv64 HiFive Unmatched board:
  - Minimal, Client and Server release & fastdebug build OK.
  - Passed tier1-tier4 tests (release build).
  - Passed jtreg tests under test/jdk/java/lang/Thread/virtual with extra JVM options: -XX:+VerifyContinuations -XX:+VerifyStack (fastdebug build).
  - Performed benchmark tests like Dacapo, SPECjvm2008, SPECjbb2015, etc. to make sure no performance regression are introduced (release build).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286301](https://bugs.openjdk.org/browse/JDK-8286301): Port JEP 425 to RISC-V

### Reviewers
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author) ⚠️ Review applies to [06302f9d](https://git.openjdk.org/jdk/pull/10917/files/06302f9d5e9e852554b70947a2d7eb63fdea69ec)
 * [Xiaolin Zheng](https://openjdk.org/census#xlinzheng) (@zhengxiaolinX - no project role) ⚠️ Review applies to [06302f9d](https://git.openjdk.org/jdk/pull/10917/files/06302f9d5e9e852554b70947a2d7eb63fdea69ec)
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author) ⚠️ Review applies to [36df84f2](https://git.openjdk.org/jdk/pull/10917/files/36df84f2c3b1771bd60f040f834cc85628f53f22)
 * [Jie Fu](https://openjdk.org/census#jiefu) (@DamonFool - **Reviewer**) ⚠️ Review applies to [850c8958](https://git.openjdk.org/jdk/pull/10917/files/850c8958fe43f913b34593f24452302c360f669a)
 * [Richard Reingruber](https://openjdk.org/census#rrich) (@reinrich - **Reviewer**)

### Contributors
 * Xiaolin Zheng `<xlinzheng@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10917/head:pull/10917` \
`$ git checkout pull/10917`

Update a local copy of the PR: \
`$ git checkout pull/10917` \
`$ git pull https://git.openjdk.org/jdk pull/10917/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10917`

View PR using the GUI difftool: \
`$ git pr show -t 10917`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10917.diff">https://git.openjdk.org/jdk/pull/10917.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:1236]: aarch64 has some assertions here, do we need this?
```cpp
#ifdef ASSERT
    if (method->is_continuation_enter_intrinsic()) {
      assert(interpreted_entry_offset != -1, "Must be set");
      assert(exception_offset != -1,         "Must be set");
    } else {
      assert(interpreted_entry_offset == -1, "Must be unset");
      assert(exception_offset == -1,         "Must be unset");
    }
    assert(frame_complete != -1,    "Must be set");
    assert(stack_slots != -1,       "Must be set");
    assert(vep_offset != -1,        "Must be set");
#endif
```

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:3879]: maybe R9 -> x9?

**Review by feilongjiang — COMMENTED** [review]
Thanks for the great work! I have reviewed `cpu/riscv` part, and here are my comments.

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:979]: Maybe I can help to answer this question.

This is an RVC-related change.
We want the `call site` itself to be aligned, for RVC can make it 2-byte aligned.
```
[code seg]
    ...
    jal <trampoline start addr>      <---   we are here, and want to force aligning this call site.
    ...


[stub seg]
    ...
<trampoline start addr:>
    auipc
    ld
    jalr
    [64-bit real address]    <-  this is certainly aligned to 8, as you've mentioned.
...
```

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (dean-long) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: Are there any ports where (stack_frame_bottom - stack_frame_top) != fsize?  It would be nice if we could use (stack_frame_bottom - stack_frame_top) for fsize and remove the platform-specific computation using frame::metadata_words above.

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:979]: Looks reasonable. As you mentioned that it‘s an RVC-related change, do we need this alignment when `UseRVC` is disabled?

File: src/hotspot/cpu/riscv/frame_riscv.inline.hpp
Comment (RealFYang) [reply src/hotspot/cpu/riscv/frame_riscv.inline.hpp:388]: Let's keep it there as Foreign-API RISC-V port is almost ready there.

File: src/hotspot/cpu/riscv/riscv.ad
Comment (RealFYang) [reply src/hotspot/cpu/riscv/riscv.ad:2461]: Note that increment/decrement may clobber 't1' register in certain cases which will conflict with the "flag" which aliases 't1' here. It will be safer to use the 'tmp' register here in this context.

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
Comment (RealFYang) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:1232]: Good catch. I think I missed this issue: "8293654: Improve SharedRuntime handling of continuation helper out-arguments". I have just added handling for RISC-V. I think this is kind of refactoring work and won't affect basic functionality. Thanks.

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
Comment (RealFYang) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:979]: This is necessary to support for RISC-V RVC extension whether it is turned on or off. And I think we are ready to enable use of RVC by default when available soon.

Comment (feilongjiang) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:979]: I see. If `UseRVC` is disabled, it will always be `NativeInstruction::instruction_size` aligned.

**Review by feilongjiang — APPROVED** [review]
Change looks good, thanks.

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
Comment (RealFYang) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:979]: That's right. Thanks again! :-)

**Review by feilongjiang — APPROVED** [review]

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (RealFYang) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: I think aarch64 and riscv is different from x86 here due to possible padding in the frame[1].
So if we modify this assertion like:
```
diff --git a/src/hotspot/share/runtime/continuationFreezeThaw.cpp b/src/hotspot/share/runtime/continuationFreezeThaw.cpp
index 2ef48618ccb..c8e88b67f94 100644
--- a/src/hotspot/share/runtime/continuationFreezeThaw.cpp
+++ b/src/hotspot/share/runtime/continuationFreezeThaw.cpp
@@ -1050,7 +1050,7 @@ NOINLINE freeze_result FreezeBase::recurse_freeze_interpreted_frame(frame& f, fr
   const int fsize = f.fp() + frame::metadata_words + locals - stack_frame_top;

   intptr_t* const stack_frame_bottom = ContinuationHelper::InterpretedFrame::frame_bottom(f);
-  assert(stack_frame_bottom - stack_frame_top >= fsize, ""); // == on x86
+  assert(stack_frame_bottom - stack_frame_top == fsize, ""); // == on x86

   DEBUG_ONLY(verify_frame_top(f, stack_frame_top));
```

Then we will trigger assertion failure on linux-aarch64:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/realfyang/openjdk-jdk/src/hotspot/share/runtime/continuationFreezeThaw.cpp
:1053), pid=2680946, tid=2680964
#  Error: assert(stack_frame_bottom - stack_frame_top == fsize) failed
#
# JRE version: OpenJDK Runtime Environment (20.0) (slowdebug build 20-internal-adhoc.realfyang.open
jdk-jdk)
# Java VM: OpenJDK 64-Bit Server VM (slowdebug 20-internal-adhoc.realfyang.openjdk-jdk, mixed mode,
 sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# V  [libjvm.so+0x865c98]  FreezeBase::recurse_freeze_interpreted_frame(frame&, frame&, int, bool)+
0xb8
#
```

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/frame_aarch64.hpp#L62

**Review by zhengxiaolinX — APPROVED** [review]
Nice overall. Please let me approve this for you :-)

**Review by yadongw — APPROVED** [review]
lgtm

**Comment by RealFYang** [comment]
Thanks all for looking at this non-trivial change! @shipilev : Want to take a look?

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (dean-long) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: OK, skipping the assert for now.  Regarding `NOT_RISCV64(+ frame::metadata_words)` @reinrich just posted the PR for the PPC64 port (see JDK-8286302), which introduces metadata_words_at_top/metadata_words_at_bottom.  I'm not sure what values RISC-V would use for these new constants, but merging with the PPC64 changes might allow a platform-independent setting of fsize here.

Comment (RealFYang) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: Hi, I went through the PPC64 changes to shared code and I think for RISC-V we should define metadata_words_at_top and metadata_words_at_bottom to 0 and 2 (which equals metadata_words as defined by this PR) respectively. The PPC64 changes will help eliminate the RISC-V-specific change made in stackChunkOopDesc::is_usable_in_chunk in file src/hotspot/share/oops/stackChunkOop.inline.hpp. But we might still need a RISC-V platform-dependent change here due to the particularity of the RISC-V frame structure. Thanks.

**Review by DamonFool — APPROVED** [review]
The shared code change looks good to me.

I'm not sure if it's possible to eliminate all the platform-dependent code.
But the current version looks good enough to me.
Thanks.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (reinrich) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: The value for `fsize` as it is calculated is correct on PPC64 but only because errors compensate. The calculation makes assumptions about the position of fp which is not specified.

This is the assumed layout:

```
  :                 :
  :                 :
  |                 |
  |-----------------|
  |                 |
  | locals array    |
  |                 |<- callers_SP
  ===================
  |                 |
  | metadata@bottom |
  |                 |<- FP
  |-----------------|
  |                 |
  |                 |
  |                 |
  |                 |
  |                 |<- SP
  ===================
```

Here the metadata lies outside [FP, SP] and needs to be added to FP-SP just like the size of the locals array.

This coincidentally matches the layout on PPC64:

```
  :                 :
  :                 :
  |                 |
  |-----------------|
  |                 |
  | locals array    |
  |                 |
  |-----------------|
  |                 |
  | metadata@top    |
  |                 |<- FP / callers_SP
  ===================
  |                 |
  |                 |
  |                 |
  |                 |
  |-----------------|
  |                 |
  | metadata@top    |
  |                 |<- SP
  ===================

```

I assume the layout on Risc-V is like this:

```
  :                 :
  :                 :
  |                 |
  |-----------------|
  |                 |
  | locals array    |
  |                 |<- FP / callers_SP
  ===================
  |                 |
  | metadata@bottom |
  |                 |
  |-----------------|
  |                 |
  |                 |
  |                 |
  |                 |
  |                 |<- SP
  ===================
```
Here the metadata is included in [FP, SP]


We could change the line to

```c++
  const int fsize = ContinuationHelper::InterpretedFrame::callers_sp(f) + frame::metadata_words_at_top + locals - stack_frame_top;
```

with a platform dependent implementation of `callers_sp()`

Comment (RealFYang) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: @reinrich : Yes, you are right about the frame layout of RISC-V. And your proposed solution looks reasonable to me. Since that will need testing accross all platforms, maybe we can do that as a separate PR later on after this is merged. The RISCV testing platform is relatively slow and this PR has been tested for a long time. Thanks.

Comment (reinrich) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: @RealFYang I'm ok with leaving this for later. The change is actually rather small. I've added it to the PPC64 port with https://github.com/openjdk/jdk/pull/10961/commits/0d12b0577d8314ed1d571d9cac65fa8866939180
Quick tests succeeded on X86, AARCH64 and PPC64.

File: src/hotspot/share/runtime/sharedRuntime.cpp
Comment (reinrich) [inline src/hotspot/share/runtime/sharedRuntime.cpp:3123]: Can we use the maximum over all platforms? This is a temporary buffer anyway. The generated code, constants, etc. is copied to the nmethod. See https://github.com/openjdk/jdk/blob/8ee0f7d5982d95674cfc1b217dbabaeafefbc8f1/src/hotspot/share/code/nmethod.cpp#L659-L660

**Review by reinrich — CHANGES_REQUESTED** [review]
Shared changes look good.
Maybe we can get rid of the RISCV64 dependency when allocating the CodeBuffer for the enter intrinsic.

Thanks, Richard.

File: src/hotspot/share/oops/stackChunkOop.inline.hpp
Comment (reinrich) [reply src/hotspot/share/oops/stackChunkOop.inline.hpp:139]: Also `frame::sender_sp_offset` is a platform detail that cannot be used in shared code. Since `metadata_words = sender_sp_offset` on other platforms I'd suggest you replace `frame::sender_sp_offset` with `frame::metadata_words`. Probably this has been forgotten.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp
Comment (RealFYang) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp:1053]: @reinrich : That's great to hear. Thank you. After this PR is merged, I can help test your changes on RISC-V after you rebase.

File: src/hotspot/share/oops/stackChunkOop.inline.hpp
Comment (RealFYang) [reply src/hotspot/share/oops/stackChunkOop.inline.hpp]: Yes, that makes sence. Fixed.

File: src/hotspot/share/runtime/sharedRuntime.cpp
Comment (RealFYang) [reply src/hotspot/share/runtime/sharedRuntime.cpp]: OK, I have set the default size to 192 for all platforms.

File: src/hotspot/share/oops/stackChunkOop.inline.hpp
Comment (reinrich) [inline src/hotspot/share/oops/stackChunkOop.inline.hpp:139]: This looks platform independent now, doesn't it? I think the cpp conditional can be removed. You can leave it also to the PPC64 port as I'll touch that line again. As you like.

**Review by reinrich — APPROVED** [review]
Looks good, thanks!
Richard.

File: src/hotspot/share/oops/stackChunkOop.inline.hpp
Comment (RealFYang) [reply src/hotspot/share/oops/stackChunkOop.inline.hpp:139]: I think I will leave it to the PR for PPC64. Thanks.

**Comment by RealFYang** [comment]
> The shared code change looks good to me.
> 
> I'm not sure if it's possible to eliminate all the platform-dependent code. But the current version looks good enough to me. Thanks.

Thanks for looking at this. I think we have enough reviews for both the shared code and riscv-specific changes now.
Let's proceed so that this won't block reviewing of PR for the loom ppc64 port.

/integrate

**Pull request closed without merging**
